### PR TITLE
Implemented BookMetaMock::setGeneration

### PR DESCRIPTION
**Pull request by svaningelgem**
# Description
I needed "setGeneration" for BookMeta, so implemented this. Also saw there was no test, so added a rudimentary one for the things that I added.

# Checklist
The following items should be checked before the pull request can be merged.
- [x] Code follows existing style.
- [x] Unit tests added (if applicable).

### CODE REVIEW

File: src/test/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMockTest.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		assertEquals(meta.getPageCount(), 0);
+
+		assertFalse(meta.hasAuthor());
+		assertFalse(meta.hasTitle());
+		assertFalse(meta.hasGeneration());
+		assertFalse(meta.hasPages());
+	}
+
+	@Test
+	void constructor_Clone_CopiesValues()
```
Comment (Thorinwasher) [inline src/test/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMockTest.java:54]: There's two ways to clone BookMetaMock, one is through the constructor (which is covered here) and another is in the Object#clone() method. Although it looks like that will work here anyhow, I think it's always a good idea to add that test

Another thing (which is an general issue not thought that much of in Mockbukkit to be real here) is that Paper almost always deep clones the objects; all mutable fields of the object has to be cloned as well (recursively). In this case it's fine implementation wise, but feel free to check if the mutable field `pages` changes reference but keeps the same contents between the two copies.

**Review by Thorinwasher — CHANGES_REQUESTED** [review]

File: src/test/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMockTest.java (hunk 1 above)
Comment (svaningelgem) [reply src/test/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMockTest.java:54]: Ok, done.

File: src/main/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMock.java [hunk 2]
```diff
@@ -250,14 +252,10 @@ public void setPages(String... pages)
...
 		this.pages.clear();
-		Iterator<String> var2 = pages.iterator();
-
-		while (var2.hasNext())
-		{
-			String page = var2.next();
-			this.addPage(page);
-		}
 
+        for (String page : pages) {
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMock.java]: This causes a ConcurrentModificationException, no? We should have tests for all methods here, whether in this PR or another.

**Review by Insprill — CHANGES_REQUESTED** [review]

File: src/main/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMock.java (hunk 2 above)
Comment (svaningelgem) [reply src/main/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMock.java]: IntelliJ gave me that hint and I thought it made sense.
It shouldn't modify anything?

```
        for (String page : pages) {
            this.addPage(page);
        }
```

And in `addPage` you just overwrite the string if it's too long, and add it into the map. So nothing being modified in there as I see? (but my knowledge of Java is lacking, so I am probably wrong)

Comment (Insprill) [reply src/main/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMock.java]: Ahh that's the list passed into the method, not the one in the class; that naming isn't confusing at all.

It should be fine then, but should still have tests.

Comment (svaningelgem) [reply src/main/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMock.java]: Can you suggest which additional tests would be needed?

Comment (Thorinwasher) [reply src/main/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMock.java]: Here's an example of methods that can still be tested.
- hashCode
- equals (this should actually fail with the current implementation)
- hasTitle
- hasAuthor
- getPages (should probably return an copied immutable list, but that's for a different pull request)

Comment (svaningelgem) [reply src/main/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMock.java]: Ok @Thorinwasher : I implemented the requested tests.

File: src/test/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMockTest.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+		List<String> pages = new ArrayList<String>();
+		pages.add("Page1");
+		pages.add("Page2");
+
+		meta.setPages(pages);
+		assertEquals(pages, meta.getPages());
+
+		pages.set(0, "Dummy");
+		assertNotEquals(pages, meta.getPages()); // testing that the `pages`' array modifications is not reflected in the book.
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMockTest.java]: It'd be clearer here to just do `assertNotSame(pages, meta.getPages())` which performs a reference equality check.

**Review by Insprill — DISMISSED** [review]

File: src/test/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMockTest.java (hunk 3 above)
Comment (svaningelgem) [reply src/test/java/be/seeseemelk/mockbukkit/inventory/meta/BookMetaMockTest.java]: Ok, didn't know that one.

I changed it now to:
```java
		meta.setPages(pages);
		assertNotSame(pages, meta.getPages());
		assertEquals(pages, meta.getPages());

		pages.set(0, "Dummy");
		assertNotSame(pages, meta.getPages()); // testing that the `pages`' array modifications is not reflected in the book.
		assertNotEquals(pages, meta.getPages()); // testing that the `pages`' array modifications is not reflected in the book.
```

ie: checking that they are NOT the same, but ARE equal.
And secondly, that they're still not the same, but now they're not equal too.

**Review by Thorinwasher — CHANGES_REQUESTED** [review]
Almost there; found one bugg though:

You need to also check if the generation is the same for the two objects to discern if they are equal to each other in the BookMetaMock#equals(Object) method. (and a good idea to add generation to the hashcode as well)

**Comment by svaningelgem** [comment]
Thanks for the review! I completely missed that. Implemented now.

**Review by Thorinwasher — APPROVED** [review]

**Pull request merged**